### PR TITLE
Feat/lmstudio integration

### LINKED ISSUE #13556: [Feature Request]: Add LMStudio support
### Feature Description

Add a way to connect LMStudio with LlamaIndex.

### Reason

Add direct implementation for LMStudio

### Value of Feature

This features enables loading models on LMStudio and allows LlamaIndex index to send request to the API provided by LMStudio.

### PR DESCRIPTION
**Pull request by shivamklr**
# Description

Add Support for LM Studio
Resolves #13556

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [x] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new notebook (that tests end-to-end)

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/llms/llama-index-llms-lmstudio/llama_index/llms/lmstudio/base.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+                    **(
+                        message.additional_kwargs
+                        if message.additional_kwargs is not None
+                        else {}
+                    ),
+                }
+                for message in messages
+            ],
+            "options": self._model_kwargs,
+            "stream": False,
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-lmstudio/llama_index/llms/lmstudio/base.py:108]: stream is hardcoded to false here (I think we want to pass this in?)

File: llama-index-integrations/llms/llama-index-llms-lmstudio/llama_index/llms/lmstudio/base.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+    timeout: float = Field(
+        default=120, description=("The timeout to use in seconds."), gte=0
+    )
+
+    additional_kwargs: Dict[str, Any] = Field(
+        default_factory=dict, description=("Additional kwargs to pass to the model.")
+    )
+
+    def __create_payload_from_messages(
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-lmstudio/llama_index/llms/lmstudio/base.py:90]: nit: lets just use single under-score for private methods

**Review by logan-markewich — APPROVED** [review]

File: llama-index-integrations/llms/llama-index-llms-lmstudio/llama_index/llms/lmstudio/base.py (hunk 2 above)
Comment (shivamklr) [reply llama-index-integrations/llms/llama-index-llms-lmstudio/llama_index/llms/lmstudio/base.py]: 👍 Fixed the typo

File: llama-index-integrations/llms/llama-index-llms-lmstudio/llama_index/llms/lmstudio/base.py (hunk 1 above)
Comment (shivamklr) [reply llama-index-integrations/llms/llama-index-llms-lmstudio/llama_index/llms/lmstudio/base.py:108]: It is not actually hardcoded. We can override the value by passing an argument in the function. In the returned dictionary, **kwargs will include all keyword arguments that are not explicitly listed in the function signature. If any of these keyword arguments match the keys in the dictionary, they will override the hardcoded values.

**Comment by shivamklr** [comment]
@logan-markewich Let me know if we have additional requirements before the merge into main.

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**